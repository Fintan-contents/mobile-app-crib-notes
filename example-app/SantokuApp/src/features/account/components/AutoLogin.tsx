import {handleErrorWithAlert} from 'bases/core/errors/handleErrorWithAlert';
import {resolveErrorMessage} from 'bases/message/resolveErrorMessage';
import React, {useEffect} from 'react';

import {useIsLoggedIn} from '../client-states/useIsLoggedIn';
import {isUnauthorizedError} from '../errors/UnauthorizedError';
import {useAuthCommands} from '../services/auth/useAuthCommands';

export const AutoLogin: React.FC<React.PropsWithChildren> = ({children}) => {
  const [isLoggedIn] = useIsLoggedIn();
  const {autoLogin} = useAuthCommands();
  useEffect(() => {
    autoLogin().catch(e => {
      if (!isUnauthorizedError(e)) {
        const {title, message} = resolveErrorMessage(e);
        handleErrorWithAlert(e, title, message);
      }
    });
  }, [autoLogin]);
  // オートログイン時にエラーが発生した場合は、isLoggedInには何も値が設定されない
  // isLoggedInに値が設定されたタイミングでオートログインの処理完了とみなす
  if (isLoggedIn == null) {
    return null;
  }
  return <>{children}</>;
};
