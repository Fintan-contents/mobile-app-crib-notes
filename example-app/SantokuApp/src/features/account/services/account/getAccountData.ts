import {getAccountsMe, getAccountsMeTerms} from 'features/backend/apis/account/account';
import {getTerms} from 'features/backend/apis/terms/terms';

export const getAccountData = async () => {
  const account = (await getAccountsMe()).data;
  const termsAgreementStatus = (await getAccountsMeTerms()).data;
  let termsOfService;
  if (!termsAgreementStatus.hasAgreed) {
    termsOfService = (await getTerms()).data;
  }
  return {
    account,
    terms: {termsAgreementStatus, termsOfService},
  };
};
