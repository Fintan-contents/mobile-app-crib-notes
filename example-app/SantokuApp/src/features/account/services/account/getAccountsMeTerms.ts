// TODO: APIが完成するまでの暫定処理
//       ローカルストレージに保存されているTermsOfServiceAgreementStatusを返却する。
//       ただし、利用規約に同意後3日経過した場合は未同意状態とする
import AsyncStorage from '@react-native-async-storage/async-storage';
import {AxiosResponse} from 'axios';
import {TermsOfServiceAgreementStatus} from 'features/backend/apis/model';

import {TermsOfServiceAgreementStatusWithCreatedAt} from '../../types/TermsOfServiceAgreementStatusWithCreatedAt';

const sleep = () => new Promise(resolve => setTimeout(resolve, 100));
const axiosGetOkResponse = {status: 200, statusText: 'OK', config: {}, request: {}, headers: {}};
export const getAccountsMeTerms = async (): Promise<AxiosResponse<TermsOfServiceAgreementStatus>> => {
  await sleep();
  const storedTermsStatus = await AsyncStorage.getItem('TermsOfServiceAgreementStatus');
  if (storedTermsStatus) {
    const parsedTermsStatus = JSON.parse(storedTermsStatus) as TermsOfServiceAgreementStatusWithCreatedAt;
    const createdAt = new Date(parsedTermsStatus.createdAt);
    const date = new Date();
    date.setDate(date.getDate() - 3);
    if (date < createdAt) {
      return {data: parsedTermsStatus, ...axiosGetOkResponse};
    }
  }

  return {
    data: {
      hasAgreed: false,
      agreedVersion: '1.0.0',
    },
    ...axiosGetOkResponse,
  };
};
