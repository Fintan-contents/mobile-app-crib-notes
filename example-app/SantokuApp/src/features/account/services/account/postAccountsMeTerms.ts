import AsyncStorage from '@react-native-async-storage/async-storage';
import {TermsOfServiceAgreementStatus} from 'features/backend/apis/model';

// TODO: APIが完成するまでの暫定処理
//       利用規約の同意状態をローカルストレージに保存する。
const axiosPostOkResponse = {status: 201, statusText: 'Created', config: {}, request: {}, headers: {}};
const sleep = () => new Promise(resolve => setTimeout(resolve, 100));
export const postAccountsMeTerms = async (data: TermsOfServiceAgreementStatus) => {
  await sleep();
  await AsyncStorage.setItem(
    'TermsOfServiceAgreementStatus',
    JSON.stringify({
      ...data,
      createdAt: new Date(),
    }),
  );
  return {
    data,
    ...axiosPostOkResponse,
  };
};
