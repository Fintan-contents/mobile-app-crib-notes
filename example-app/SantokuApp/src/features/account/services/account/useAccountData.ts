import {useQuery} from 'react-query';

import {getAccountData} from './getAccountData';

const queryKey = ['account', 'accountData'];

export const useAccountData = (args: {enabled?: boolean; meta?: {disableGlobalErrorHandler?: boolean}} = {}) => {
  const query = useQuery(queryKey, getAccountData, {enabled: args.enabled, meta: args.meta, staleTime: Infinity});
  return {...query, accountData: query.data};
};
