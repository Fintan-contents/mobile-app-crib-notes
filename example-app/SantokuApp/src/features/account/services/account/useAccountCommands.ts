import {useMutation, useQueryClient} from 'react-query';

import {postAccountsMeTerms} from './postAccountsMeTerms';

export const useAccountCommands = () => {
  const queryClient = useQueryClient();

  /**
   * サービス利用規約に同意します。
   */
  const agreeTermsMutation = useMutation(postAccountsMeTerms, {
    onSuccess: () => {
      queryClient.resetQueries(['account', 'accountData']).catch(() => {});
    },
  });

  return {
    agreeTerms: agreeTermsMutation.mutateAsync,
    isaAgreeingTerms: agreeTermsMutation.isLoading,
    isProcessing: agreeTermsMutation.isLoading,
  };
};
