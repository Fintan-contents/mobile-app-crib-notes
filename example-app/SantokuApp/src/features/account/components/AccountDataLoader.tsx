import {sendErrorLog} from 'bases/logging/sendErrorLog';
import {resolveErrorMessage} from 'bases/message/resolveErrorMessage';
import React, {useEffect, useRef} from 'react';
import {Alert} from 'react-native';

import {useIsLoggedIn} from '../client-states/useIsLoggedIn';
import {useAccountData} from '../services/account/useAccountData';

export const AccountDataLoader: React.FC = ({children}) => {
  const [isLoggedIn] = useIsLoggedIn();
  const isAutoLoggedIn = useRef(isLoggedIn).current;
  const {isLoading, isLoadingError, errorUpdateCount, error} = useAccountData({
    enabled: isLoggedIn,
    meta: {disableGlobalErrorHandler: true},
  });

  useEffect(() => {
    // オートログイン済で初回のデータ取得時にエラーが発生した場合は
    if (isAutoLoggedIn && isLoadingError) {
      if (errorUpdateCount === 1) {
        const {title, message} = resolveErrorMessage(error);
        sendErrorLog(error);
        Alert.alert(title, message);
      }
    }
  }, [error, errorUpdateCount, isAutoLoggedIn, isLoadingError]);

  if (isAutoLoggedIn && (isLoading || isLoadingError)) {
    return null;
  }
  return <>{children}</>;
};
