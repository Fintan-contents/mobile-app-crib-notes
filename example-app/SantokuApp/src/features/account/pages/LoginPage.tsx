import {m} from 'bases/message/Message';
import {FilledButton} from 'bases/ui/button/FilledButton';
import {OutlinedButton} from 'bases/ui/button/OutlinedButton';
import {PasswordTextInput} from 'bases/ui/input/PasswordTextInput';
import {TextInput} from 'bases/ui/input/TextInput';
import {Spacer} from 'bases/ui/spacer/Spacer';
import {TermsOfServiceAgreementStatus} from 'features/backend/apis/model';
import {TermsAgreementOverlay} from 'features/terms/components/TermsAgreementOverlay';
import {useTerms} from 'features/terms/services/useTerms';
import React, {useCallback} from 'react';
import {Alert, StyleSheet, View} from 'react-native';

import {isUnauthorizedError} from '../errors/UnauthorizedError';
import {LoginFormValues, useLoginForm} from '../forms/useLoginForm';
import {useAuthCommands} from '../services/auth/useAuthCommands';

export type LoginPageProps = {
  navigateToCreateAccount: (termsOfServiceAgreementStatus: TermsOfServiceAgreementStatus) => void;
};

export const LoginPage: React.FC<LoginPageProps> = ({navigateToCreateAccount}) => {
  const {login, isLoggingIn} = useAuthCommands();
  const onSubmit = useCallback(
    async (values: LoginFormValues) => {
      try {
        await login({accountId: values.accountId, password: values.password});
      } catch (e) {
        if (isUnauthorizedError(e)) {
          Alert.alert(m('ログイン失敗'), m('アカウントIDまたはパスワードに\n間違いがあります。'));
        }
      }
    },
    [login],
  );
  const {form, clearAccountId, clearPassword} = useLoginForm({
    onSubmit,
  });
  const {termsOfService, isFetchedTerms} = useTerms();

  const onCreateAccount = useCallback(() => {
    TermsAgreementOverlay.show({
      // 利用規約を取得できていない場合はボタンを非活性にしているので、ここでは必ず存在する想定
      termsOfService: termsOfService!.data,
      exitingCallbackOnAgreed: (termsOfServiceAgreementStatus: TermsOfServiceAgreementStatus) => {
        navigateToCreateAccount(termsOfServiceAgreementStatus);
      },
      dismissible: true,
    });
  }, [navigateToCreateAccount, termsOfService]);

  return (
    <View style={styles.container} testID="Login">
      <TextInput
        label={m('アカウントID')}
        value={form.values.accountId}
        onChangeText={form.handleChange('accountId')}
        showClearButton={!!form.values.accountId}
        onClearButtonPress={clearAccountId}
        errorMessage={form.errors.accountId}
      />
      <Spacer heightRatio={0.03} />
      <PasswordTextInput
        label={m('パスワード')}
        value={form.values.password}
        onChangeText={form.handleChange('password')}
        showClearButton={!!form.values.password}
        onClearButtonPress={clearPassword}
        errorMessage={form.errors.password}
      />
      <Spacer heightRatio={0.05} />
      <View style={{flexDirection: 'row', justifyContent: 'center'}}>
        <OutlinedButton title={m('新規登録')} onPress={onCreateAccount} disabled={!isFetchedTerms} />
        <Spacer widthRatio={0.1} />
        <FilledButton
          title={m('ログイン')}
          onPress={form.submitForm}
          loading={isLoggingIn}
          disabled={!isFetchedTerms}
        />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    paddingHorizontal: 30,
  },
});
