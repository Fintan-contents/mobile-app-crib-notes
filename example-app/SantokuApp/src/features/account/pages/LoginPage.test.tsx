import {render, screen, waitFor} from '@testing-library/react-native';
import {BundledMessagesLoader} from 'bases/message/BundledMessageLoader';
import {loadMessages} from 'bases/message/Message';
import {Snackbar} from 'bases/ui/snackbar/Snackbar';
import {AppThemeProvider} from 'bases/ui/theme/AppThemeProvider';
import {enhanceValidator} from 'bases/validator';
import {useGetAccountsMe, useGetAccountsMeTerms, usePostAccountsMeTerms} from 'features/backend/apis/account/account';
import {useGetTerms} from 'features/backend/apis/terms/terms';
import {TermsAgreementOverlay} from 'features/terms/components/TermsAgreementOverlay';
import React from 'react';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import {QueryClient, QueryClientProvider} from 'react-query';

import {LoginPageProps} from './LoginPage';

jest.mock('features/backend/apis/account/account');
jest.mock('features/backend/apis/terms/terms');

const Wrapper: React.FC<React.PropsWithChildren> = ({children}) => {
  const queryClient = new QueryClient();
  return (
    <SafeAreaProvider>
      <AppThemeProvider>
        <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
        <TermsAgreementOverlay.Component />
        <Snackbar.Component />
      </AppThemeProvider>
    </SafeAreaProvider>
  );
};

beforeAll(async () => {
  await loadMessages(new BundledMessagesLoader());
  enhanceValidator();
});

describe('LoginScreen', () => {
  it('マウントされたときに正常にレンダリングされること', async () => {
    (useGetAccountsMe as jest.Mock).mockReturnValue({refetch: () => {}});
    (useGetAccountsMeTerms as jest.Mock).mockReturnValue({refetch: () => {}});
    (useGetTerms as jest.Mock).mockReturnValue({refetch: () => {}});
    (usePostAccountsMeTerms as jest.Mock).mockReturnValue({mutateAsync: () => {}});
    // importでLoginScreenを読み込むと、メッセージのロードが完了する前にメッセージを読み込んでしまうため、requireで取得する
    // requireした場合の型はanyとなってしまいESLintエラーが発生しますが無視します。
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const Page = require('./LoginPage').LoginPage as React.FC<React.PropsWithChildren<LoginPageProps>>;
    render(<Page navigateToCreateAccount={() => {}} />, {
      wrapper: Wrapper,
    });
    await waitFor(() => {
      expect(screen.queryByTestId('Login')).not.toBeNull();
      expect(screen).toMatchSnapshot();
    });
  });
});
