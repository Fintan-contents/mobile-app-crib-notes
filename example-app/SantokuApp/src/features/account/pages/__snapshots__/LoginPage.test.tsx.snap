// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen マウントされたときに正常にレンダリングされること 1`] = `
Array [
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 30,
      }
    }
    testID="Login"
  >
    <View
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": 0,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#212935",
              "fontSize": 15,
              "fontWeight": "normal",
            }
          }
        >
          アカウントID
        </Text>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#9e9e9e",
              "borderBottomWidth": 2,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "paddingVertical": 8,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            editable={true}
            onChangeText={[Function]}
            placeholderTextColor="#86939e"
            style={
              Object {
                "color": "#212935",
                "flex": 1,
                "fontSize": 17,
                "includeFontPadding": false,
                "lineHeight": 24,
                "minHeight": 0,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "height": 0,
              "margin": 0,
              "padding": 0,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "height": 19.2,
          "width": 0,
        }
      }
      testID="Spacer"
    />
    <View
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": 0,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#212935",
              "fontSize": 15,
              "fontWeight": "normal",
            }
          }
        >
          パスワード
        </Text>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#9e9e9e",
              "borderBottomWidth": 2,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "paddingVertical": 8,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoComplete="password"
            autoCorrect={false}
            editable={true}
            keyboardType="ascii-capable"
            onChangeText={[Function]}
            placeholderTextColor="#86939e"
            secureTextEntry={true}
            style={
              Object {
                "color": "#212935",
                "flex": 1,
                "fontSize": 17,
                "includeFontPadding": false,
                "lineHeight": 24,
                "minHeight": 0,
              }
            }
            testID="RNE__Input__text-input"
            textContentType="password"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "height": undefined,
                "justifyContent": "center",
                "marginVertical": 0,
                "paddingRight": 4,
                "paddingVertical": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="IconButtonTouchableOpacity"
              >
                <View
                  style={
                    Object {
                      "overflow": "hidden",
                    }
                  }
                >
                  <View
                    replaceTheme={[Function]}
                    style={
                      Object {
                        "width": 0,
                      }
                    }
                    updateTheme={[Function]}
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "justifyContent": "center",
                        }
                      }
                    >
                      <Text />
                    </View>
                  </View>
                </View>
              </View>
              <View
                style={
                  Object {
                    "marginRight": 5,
                  }
                }
              />
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="IconButtonTouchableOpacity"
              >
                <View
                  style={
                    Object {
                      "overflow": "hidden",
                    }
                  }
                >
                  <View
                    replaceTheme={[Function]}
                    updateTheme={[Function]}
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "justifyContent": "center",
                        }
                      }
                    >
                      <Text />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "height": 0,
              "margin": 0,
              "padding": 0,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "height": 32,
          "width": 0,
        }
      }
      testID="Spacer"
    />
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            Object {
              "borderRadius": 8,
              "width": 90,
            },
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "hsl(208, 8%, 90%)",
                "borderColor": "#3a6be8",
                "borderRadius": 8,
                "borderWidth": 0,
                "flexDirection": "row",
                "height": 50,
                "justifyContent": "center",
                "padding": 8,
                "width": 90,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "hsl(208, 8%, 63%)",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              新規登録
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 0,
            "width": 32,
          }
        }
        testID="Spacer"
      />
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            Object {
              "borderRadius": 8,
              "width": 90,
            },
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "hsl(208, 8%, 90%)",
                "borderColor": "#2089dc",
                "borderRadius": 8,
                "borderWidth": 0,
                "flexDirection": "row",
                "height": 50,
                "justifyContent": "center",
                "padding": 8,
                "width": 90,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "hsl(208, 8%, 63%)",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              ログイン
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    testID="FullWindowOverlay"
  />,
  <View
    testID="FullWindowOverlay"
  />,
]
`;
